crates/matrix-sdk/src/room_preview.rs

```diff
@@ -140,21 +142,117 @@ impl RoomPreview {
+
+            let (results, _) = directory_search.results();
+
+            for room_description in results {
+                // Iterate until we find a room description with a matching room id
+                if room_description.room_id != expected_room_id {
+                    continue;
+                }
+                return Ok(RoomPreview {
+                    room_id: room_description.room_id,
+                    canonical_alias: room_description.alias,
+                    name: room_description.name,
+                    topic: room_description.topic,
+                    avatar_url: room_description.avatar_url,
+                    num_joined_members: room_description.joined_members,
+                    num_active_members: None,
+                    // Assume it's a room
+                    room_type: None,
+                    join_rule: match room_description.join_rule {
+                        PublicRoomJoinRule::Public => SpaceRoomJoinRule::Public,
+                        PublicRoomJoinRule::Knock => SpaceRoomJoinRule::Knock,
+                        PublicRoomJoinRule::_Custom(rule) => SpaceRoomJoinRule::_Custom(rule),
+                        _ => {
+                            panic!("Unexpected PublicRoomJoinRule {:?}", room_description.join_rule)
+                        }
+                    },
+                    is_world_readable: room_description.is_world_readable,
+                    state: None,
+                    is_direct: None,
+                });
+            }
+
+            Err(crate::Error::InsufficientData)
+        }
+
+        // Get either the room alias or the room id without the leading identifier char
+        let search_term = if room_or_alias_id.is_room_alias_id() {
+            Some(room_or_alias_id.as_str()[1..].to_owned())
+        } else {
+            None
+        };
+
+        // If we have no alias, filtering using a room id is impossible, so just take
+        // the first 100 results and try to find the current room #YOLO
+        let batch_size = if search_term.is_some() { 20 } else { 100 };
+
+        let futures = if via.is_empty() {
+            // Just search in the current homeserver
+            vec![search_homeserver(client.clone(), search_term, batch_size, None, room_id)]
+        } else {
+            let mut futures = Vec::new();
+            // Iterate all servers until one returns some results
+            for server in via {
+                futures.push(search_homeserver(
+                    client.clone(),
+                    search_term.clone(),
+                    batch_size,
+                    Some(server),
+                    room_id,
+                ));
+            }
+            futures
+        };
+
+        let joined_results = join_all(futures).await;
+
+        if let Some(preview) = joined_results.into_iter().flatten().next() {
+            return Ok(preview);
+        }
+
+        Err(crate::Error::InsufficientData)
```

Comment:
Let's have this function return `Result<Option<...>>`, so we can return `Ok(None)` here. It's not really an error if we couldn't find a room preview with this method.

Reply:
Yeah, that's better, thanks!